### PR TITLE
[Enhancement] PolicyBase well-defined and corresponding type_traits

### PR DESCRIPTION
**Pull request by willkill07**
**NOTE** this has been heavily updated and the text at the top is the most recent implementation

`PolicyBase` now defines different types of policies and types:

here's a proposed change that explicitly breaks apart policy, launch, and pattern. This would also make `Pattern`, `Policy` and `Launch` more friendly for user-facing logic :)

```cpp
enum class Policy {
  undefined,
  sequential,
  simd,
  openmp,
  cuda,
  cilk
};

enum class Launch {
  undefined,
  sync,
  async
};

enum class Pattern {
  undefined,
  forall,
  reduce,
  taskgraph,
};

template <
  Policy P = Policy::undefined,
  Launch L = Launch::undefined,
  Pattern Pat = Pattern::undefined>
struct PolicyBase {
  static constexpr Policy policy = P;
  static constexpr Launch launch = L;
  static constexpr Pattern pattern = Pat;
};

template <typename Inner, typename ... T>
struct WrapperPolicy : public Inner {
  using inner = Inner;
};
```

type traits exist for:

```cpp
is_<POLICY>_policy
is_<LAUNCH>_launch
is_<PATTERN>_pattern
```

creating templated policies and patterns are also easy:

```cpp
{POLICY}_for< Pattern > // openmp_for<Pattern::forall>
{PATTERN}_for< Policy > // reduce_for<Policy::cuda>
```

extracting policy/launch/patterns are now trivial:

```cpp
using Pol = RAJA::omp_parallel_for_exec;
RAJA::Policy policy = Pol::policy;
RAJA::Pattern pattern = Pol::pattern;
```

Definitions in each *policy* header file now properly extend from a corresponding policy

### CODE REVIEW

File: include/RAJA/policy/cuda/policy.hpp [hunk 1]
```diff
@@ -66,7 +68,7 @@ struct Dim3z {
 /// Segment execution policies
 ///
 
-struct cuda_exec_base {
+struct cuda_exec_base : public forall_policy {
```
Comment (DavidPoliakoff) [inline include/RAJA/policy/cuda/policy.hpp]: This is interesting. Essentially here we say that reduce (and taskgraph?) policies will never inherit from backend_exec_base policies, that is, everything which inherits from cuda_exec_base must be a forall_policy. I don't know that it's wrong, I just think it should be discussed

edit: the more I look at this, the more I think you might have the basis of a good refactoring, note how some backends have an exec_base, some (like seq) just have an exec forall policy.

**Comment by DavidPoliakoff** [comment]
Okay, think I've got this together.

As things are, the way different backends organize their policies is inconsistent. CUDA has a base execution policy which only forall's inherit from, ditto openmp (?). Sequential has no base, just a seq_exec and seq_reduce. These inconsistent semantics are a little iffy, if I want to extend CUDA foralls, I inherit from the cuda_exec_base and automatically become a correct forall_policy. If I want to enhance seq_exec (perhaps a more strong "thou shalt not parallelize" pragma) I write a new_seq_exec and explicitly call it a forall_policy. This is tedious, but I want to get it right, so I have two possible suggestions.

Under each, we have policy_base, and your pattern bases as they are, these are excellent. Under each, we also have a backend_base policy which **is not a forall**.

1) under backend_base, we have backend_forall_base, backend_reduce_base, backend_taskgraph_base which correctly inherit from forall_policy AND backend_base. Our current implementations then correctly inherit from these policies.

2) Have the backend base, then have any implementations be responsible for correct inheritance. This may feel like a strawman, but it is less constraining and I don't know that my first suggestion correctly classifies everything we want.

Nice PR, it's kinda the start of answering @trws 's old issue on having rigorously defined semantics for RAJA constructs.

**Comment by DavidPoliakoff** [comment]
Also, if we're doing this (1), adding is_cuda_policy and its ilk to the type_traits would be nice.

**Comment by willkill07** [comment]
> CUDA has a base execution policy

Yeah, and it's the only one. I am removing it as it is not used anywhere meaningful (only used in a storage class for scan tests).

My suggestion would be to define `policy_tag` for each *IMPL* we have. These tags can all be defined in `PolicyBase` and have `is_IMPL_policy` defined under `type_traits.hpp`

I think this is finally a good use case for class enums:

```cpp
enum class policy_tag {
  sequential,
  simd,
  openmp,
  cuda,
  milk
};
```

PolicyBase would get a template parameter indicating the `policy_tag` with a hidden `detail::PolicyBase` used for type traits.

I'll work on an update to this PR that conforms to this (I can always revert)

**Comment by DavidPoliakoff** [comment]
Nifty, that satisfies the constraints I want. We'd need to examine whether these impl's could ever be hierarchical, and if so whether the tag system works for us. That's not me saying "it won't," it's me saying "I don't know enough about class enums to comment."

Also, it might be wise to think of this from the type_traits side of things. Rather than make an organization and then see what we can ask of it, see what we want to ask first. Things that come to mind:

1. Is this thing a forall? (solved)
2. Is this thing CUDA'ish? (solved)
3. Is this thing asynchronous? (Not solved that I know? Should it be solved here? Is asynchrony a feature of a policy or a feature of the launching construct? Should it be that way? Wargarbl and various other "I don't really know" sounds)

Is there anything else people can think of we might want to inquire about?

**Comment by willkill07** [comment]
@DavidPoliakoff here's a snippet that allows a policy to be defined by any number of policy tags. In terms of extensibility, this is an obvious winner. The implementation of `any_of` is also quite efficient, as there is only one template nesting.

https://godbolt.org/g/dgZEGM

Let me know what you think of this. I'll have to reconsider how wrapping policies (e.g. openmp) works with this approach.

**Comment by DavidPoliakoff** [comment]
@willkill07 : this looks really neat to me, I honestly don't know enough about how well it'll play with the wrapping. I'd wait on somebody else's comments as well, but I quite like the system you've put in there. I worry about a new developer's experience with this, but honestly from the godbolt it looks pretty innocuous. Any other thoughts? @davidbeckingsale ?

**Comment by willkill07** [comment]
As it turns out, I was overthinking the wrapped policy. The following works:

```cpp
// add wrapped to policy_tag

template <typename Inner, policy_tag ... Tags>
struct PolicyWrapper : public PolicyBase<policy_tag::wrapped, Tags...> {
  using inner_policy = Inner;
};
```

**Comment by rhornung67** [comment]
@willkill07 I like this, but I have to spend more time with it and wrap my head around it to understand it better?  Would it be possible to use this policy framework and/or the policy tag enum to avoid separation of the reduce policy and loop execution policy; e.g., user declares a policy agnostic reduction object, when it gets lambda-captured the execution policy is used to create the proper PM-specific reduction object, etc.?

**Comment by DavidPoliakoff** [comment]
@rhornung67 : I don't think we get that, but it does raise an interesting point. I wonder whether we could implement reverse lookups, like "give me the/some reducer that comes with OpenMP backend." Then we could write code like
`
using backend = cuda_backend;
using forall = forall_for<backend>; //named something else plz
using reduce = reduce_for<backend>;
`

This would lead to fewer code changes to switch backends. I don't know if capture based shenanigans are possible that way though

**Comment by willkill07** [comment]
@rhornung67 yes, that's exactly why I created this PR. My next step would be to revisit reductions and combine the policy types here with RAJA::operators. This would refactor the reduction types to a common base. Then any reduction could be specified with 1). a reduce policy 2). a reduce function

**Comment by willkill07** [comment]
@DavidPoliakoff I'm not sure if you noticed, but `policy_tag` accomplishes that exactly :)

`PolicyBase<policy_tag::forall, policy_tag::openmp>` is the type for OpenMP

I suppose I could throw in a few helpers:

```cpp
template <typename Pol>
using forall_for = PolicyBase<policy_tag::forall, Pol>;
template <typename Pol>
using reduce_for = PolicyBase<policy_tag::reduce, Pol>;
```

Likewise, it could be used in reverse as well -- given an execution policy, I can request a pattern:

```cpp
template <typename Pattern>
using simd_pattern = PolicyBase<policy_tag::simd, Pattern>;
template <typename Pattern>
using cuda_pattern = PolicyBase<policy_tag::cuda, Pattern>;
template <typename Pattern>
using cuda_pattern_async = PolicyBase<policy_tag::cuda, policy_tag::async, Pattern>;
```

**Comment by DavidPoliakoff** [comment]
@willkill07 : it's more than this needs, but something less ham-fisted than line 85 of 

https://godbolt.org/g/FwQD4C

would be cool

**Comment by DavidPoliakoff** [comment]
@willkill07 : also fyi I think this is the coolest PR for RAJA since the Iterable refactoring, code will get a lot more generic if these semantics work. Really nice stuff.

**Comment by DavidPoliakoff** [comment]
I make something of a mess of mocking up reducers, but the end result on line 107 of this

https://godbolt.org/g/OvoscT

Fills me with joy, and if we could find a way to do it that people were happy with, I'd like it. @ajkunen  , you could do really silly things with these kinds of constructs, though I think you've hidden the need for them in your own code

**Comment by willkill07** [comment]
@DavidPoliakoff and everyone ... the proposed method above wasn't quite robust enough.

here's a proposed change that explicitly breaks apart policy, launch, and pattern

```cpp
enum class Policy {
  undefined,
  sequential,
  simd,
  openmp,
  cuda,
  cilk
};

enum class Launch {
  undefined,
  sync,
  async
};

enum class Pattern {
  undefined,
  forall,
  reduce,
  scan,
  taskgraph,
};

template <
  Policy P = Policy::undefined,
  Launch L = Launch::undefined,
  Pattern Pat = Pattern::undefined>
struct PolicyBase {
  static constexpr Policy policy = P;
  static constexpr Launch launch = L;
  static constexpr Pattern pattern = Pat;
};

template <typename Inner, typename ... T>
struct WrapperPolicy : public Inner {
  using inner = Inner;
};

template <
  Policy P, Launch L, Pattern Pat,
  typename ... Tags>
struct WrapperPolicy<PolicyBase<P,L,Pat>, Tags...> : public PolicyBase<P,L,Pat> {
  using inner = PolicyBase<P,L,Pat>;
};
```

type traits exist for:

```cpp
is_<POLICY>_policy
is_<LAUNCH>_launch
is_<PATTERN>_pattern
```

extracting policy/launch/patterns are now trivial:

```cpp
using Pol = RAJA::omp_parallel_for_exec;
RAJA::Policy policy = Pol::policy;
RAJA::Pattern pattern = Pol::pattern;
```

https://godbolt.org/g/XXR7oP

**Comment by DavidPoliakoff** [comment]
Bummer! I'll have to hack on that tomorrow, I know it has to work (note: this typically means that it doesn't actually work). I really like what you have, with the possible caveat that "sync/async" isn't enough to describe _all_ of the launch characteristics we might want, that critique is blunted by the fact that we don't have to describe every kind of asynchrony we can think of in this way.

And I know we're not writing C++17, but the ability in a user code to do something like
```
template<typename policy_type>
void do_things(){
   if constexpr(is_cuda_policy<policy_type>::value){
       //special handling for memory movement code
   }
}
```

is really neat

**Comment by willkill07** [comment]
> with the possible caveat that "sync/async" isn't enough to describe all of the launch characteristics we may want

which is one reason why it's in an enum -- we can always add to it. Just like we will have new policies over time and additional patterns.

**Comment by DavidPoliakoff** [comment]
Yeah, my only worry there (and I may just be losing my mind here) is whether we still had block sizes on our CUDA exec policies, whether it was cuda_exec or cuda_exec<512> and whether we can support that... really hope I'm wrong there. But generally this lets us describe discrete categories/tags quite well

**Comment by willkill07** [comment]
@DavidPoliakoff our built-in types will remain the same -- our logic will remain the same.

What changes is the behavior when a user wants a pattern from a policy (or vice versa) -- at that point we will have to rely on a good-enough default.

Given the programmatic nature of such a feature (or bug depending on your viewpoint), it may be best to mark such constructs as experimental

**Comment by DavidPoliakoff** [comment]
Right! I had forgotten, this isn't an overwrite, this is just changing the structure of what we're inheriting from. Slow moment, thanks for the patience!

**Comment by trws** [comment]
> @willkill07 I like this, but I have to spend more time with it and wrap my head around it to understand it better? Would it be possible to use this policy framework and/or the policy tag enum to avoid separation of the reduce policy and loop execution policy; e.g., user declares a policy agnostic reduction object, when it gets lambda-captured the execution policy is used to create the proper PM-specific reduction object, etc.?

I'm just circling back around to RAJA after ERC stuff (yay...), and I realize this is a bit off-topic, but this is something I'd like to find a way to do @rhornung67, we just can't do it with capture because capture can't change the type of the reducer.  If we find an interface to something like the transform_reduce in #237 that people find palatable we can do that, we can also make reductions faster, less restrictive, and start using compiler supported reduction operations internally from models like OpenMP.  It would also be possible to write a reducer that works with agency, which is otherwise not feasible right now.

**Comment by trws** [comment]
Actually on this PR, I *love* the rework to this @willkill07, it's more than I was even envisioning when I put in the original PolicyBase (badly, I admit, sigh...).   Thanks a lot for taking the time to do this properly, it goes a long way toward making it possible to write up a real concept specification for RAJA policies and foralls, which would be really awesome for everyone IMO.

**Comment by willkill07** [comment]
**OFF TOPIC COMMENT**

> I'm just circling back around to RAJA after ERC stuff (yay...), and I realize this is a bit off-topic, but this is something I'd like to find a way to do @rhornung67, we just can't do it with capture because capture can't change the type of the reducer.

@trws -- the C++14 expanded capture would allow us to do something like:

```cpp
using Pol = RAJA::omp_parallel_for_exec;
Reduce<RAJA::operator::min, T> min;
RAJA::forall<Pol>(RAJA::RangeSegment{0, 100}, [=, min = RAJA::use<Pol>(min)] (int x) {
  min.reduce(array[i]);
});
```

`Reduce` would have template parameters: `template <template <typename...> class Op, typename ... Ts>` where we could specialize for single values `<T>` or index-value pairs `<int, T>`. `Reduce` would only provide a `reduce` method which expects arguments to be the same as `Ts...` and ultimately call `Op<Ts...>::operator()`.

Most importantly, `RAJA::use<Policy>` would be defined for each type of policy and do a similar setup/teardown necessary for each policy type. The limitation with this approach is reductions would need to happen in the same algorithmic way for each policies (tree or left-fold reductions).

Limitation: each reducer would need to have an explicit move and this would require C++14. The `Reduce` construct is still adaptable within C++11 and would work well with a refactored reduction interface that takes advantage of expanded `PolicyBase`. We'd need to add a `Policy` template parameter to Reduce to have it work with existing machinery.

**Comment by DavidPoliakoff** [comment]
@willkill07 THIS IS EXACTLY WHAT I'VE BEEN WAITING FOR. Man that's cool. Compiler plugins are going to be so much fun with 14.

**Comment by DavidPoliakoff** [comment]
@willkill07 , wait, it's still whatever reducer type it was on the outside at the outside scope, so we have to make sure that value retrieval is well-formed. Not sure that's a problem, not sure it's not

**Comment by willkill07** [comment]
> it's still whatever reducer type it was on the outside at the outside scope

Yes, it would be `uses` job to ensure the result value gets updated *safely* to the parent (e.g. storing a reference/ptr to the original).

**Comment by trws** [comment]
I agree that would work @willkill07, but it still requires the policy to be explicitly specified on each reduction variable, which is what I thought @rhornung67 was requesting we avoid.  Don't get me wrong, I actually really like what the explicit renaming capture lets us do, but I'm not sure it quite covers the use-case as I was seeing it.

**Comment by DavidPoliakoff** [comment]
I still really want to hack together a "reducer_for<policy_tag::cuda,ReduceSum>::type" system, can't do it right now though. It's a lot of scut work writing specializations, but I think it would allow for changing the backend to change both reducer and forall types.

**Comment by davidbeckingsale** [comment]
Looking at this, I think it's good to merge. One quick question - do the traits work for ForallN policies? It's not a requirement, but it's definitely something to think about...

**Comment by willkill07** [comment]
@davidbeckingsale at the moment, no. It's on the back burner slowly simmering until I figure out a good type design that won't destroy compilation times.

The current `forallN` machinery relies on tags -- ideally we'd like to find a way to describe loop transformations with `PolicyTraits` as well.

**Comment by davidbeckingsale** [comment]
No problem 😄  It's definitely not a requirement to get this merged. I really like the traits you've added, and it's super useful to some stuff I'm doing internally.

**Comment by trws** [comment]
Yeah, I'd leave that machinery alone until I'm done with the foralln refactor I'm working on...  Hopefully it will actually make it easier, but either way it would be sufficiently disruptive that it's probably not useful to do it just now.

**Comment by willkill07** [comment]
Alright. This looks like a good place state shipit :shipit:

**Comment by trws** [comment]
Want to do the update from develop?

**Comment by davidbeckingsale** [comment]
This broke CUDA, removed the `cuda_exec_async` shorthand policy 😢

**Comment by davidbeckingsale** [comment]
@willkill07 there are a couple of CUDA bugs in this - the removal of `cuda_exec_async`, and the use of `is_cuda_policy` (in scan-tests) aren't working.

**Pull request merged**